### PR TITLE
Improve gl pull and gl fetch

### PR DESCRIPTION
**Pull request by sirlucjan**
A small change that will make gl fetch and gl pull similar to other gitless commands

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>

### CODE REVIEW

File: gitless/cli/gl_fetch.py [hunk 1]
```diff
@@ -21,3 +22,5 @@ def parser(subparsers, _):
 
 def main(args, repo):
     core.git_wrap('fetch', *args.fetch_args)
+    pprint.ok('Fetching from upstream succeeded')
```
Comment (goldstar611) [inline gitless/cli/gl_fetch.py]: The fetch could fail for any number of reasons which will be printed to stdout/stderr. We should check the `returncode` from the `git_wrap()` call and only print the ok message if `returncode == 0`. Without the check the user will always see a success message which may not be true.

Same applies to:
- [ ] gl_pull.py
- [ ] [gl_patch.py](https://github.com/goldstar611/gitless/pull/33/files#diff-e7155535a5cadd32eeb2bb38c5fcc31228792b6005b7b6c64fef2f2f29efd65d)

File: gitless/cli/gl_fetch.py [hunk 2]
```diff
@@ -6,6 +6,7 @@
 
 
 from ..import core
+from . import helpers, pprint
```
Comment (goldstar611) [inline gitless/cli/gl_fetch.py]: The `helpers` import is not used here. Could you remove please?

Same applies to:
- [x] gl_pull.py
- [x] [gl_patch.py](https://github.com/goldstar611/gitless/pull/33/files#diff-e7155535a5cadd32eeb2bb38c5fcc31228792b6005b7b6c64fef2f2f29efd65d)

**Review by goldstar611 — CHANGES_REQUESTED** [review]
This is a good idea and improves the ui.

- [x] Action item for me to add some smoke tests to the suite
  - [x] A basic "smoke" test that should "pass" e.g. return 0
  - [x] A basic "smoke" test that should "fail" e.g. return non-zero

File: gitless/cli/gl_fetch.py [hunk 3]
```diff
@@ -21,3 +22,5 @@ def parser(subparsers, _):
 
 def main(args, repo):
     core.git_wrap('fetch', *args.fetch_args)
+    pprint.ok('Fetching from upstream succeeded')
+    return True
```
Comment (goldstar611) [inline gitless/cli/gl_fetch.py]: I would also suggest changing `return True` to something like  `return p.returncode == 0` where `p` is the object return by `git_wrap`. I need to see if this boolean is propagated upstream too. (✔it is)

Same applies to:
- [x] gl_pull.py
- [x] [gl_patch.py](https://github.com/goldstar611/gitless/pull/33/files#diff-e7155535a5cadd32eeb2bb38c5fcc31228792b6005b7b6c64fef2f2f29efd65d)

File: gitless/cli/gl_fetch.py [hunk 4]
```diff
@@ -23,4 +23,4 @@ def parser(subparsers, _):
 def main(args, repo):
     core.git_wrap('fetch', *args.fetch_args)
```
Comment (goldstar611) [inline gitless/cli/gl_fetch.py]: add `p = ` before `core.git_wrap(....` and you should be good to go.

Comment (sirlucjan) [reply gitless/cli/gl_fetch.py]: Done. Maybe I'll need more coffee :D

**Comment by sirlucjan** [comment]
The only thing needed now is to add pprint.err in case something goes wrong.

**Comment by goldstar611** [comment]
> The only thing needed now is to add pprint.err in case something goes wrong.

It's looking good. I pushed some basic tests to your branch also.
Thanks for your contributions :)

**Comment by sirlucjan** [comment]
I will try to add prrint.err in the near future if necessary.

**Comment by goldstar611** [comment]
> I will try to add prrint.err in the near future if necessary.

I can add it for you, just let me know. I'm also happy to leave the PR open until you get around to it.

**Comment by sirlucjan** [comment]
If I could ask for a suggestion - how would you resolve this pprint.err?

**Comment by goldstar611** [comment]
> If I could ask for a suggestion - how would you resolve this pprint.err?

Since we have the variable `p` what I would do is
```python
if p.returncode == 0:
    pprint.ok('Fetching from upstream succeeded')
else:
    pprint.err("Fetching from upstream failed")
```

[Edit] Sorry I mixed single and double quotes. It looks like the style here is single quotes. :D

**Comment by sirlucjan** [comment]
> ```python
> if p.returncode == 0:
>     pprint.ok('Fetching from upstream succeeded')
> else:
>     pprint.err("Fetching from upstream failed")
> ```

It doesn't seem to be working all the way.

```Traceback (most recent call last):
  File "/usr/bin/gl", line 33, in <module>
    sys.exit(load_entry_point('gitless==0.9.16', 'console_scripts', 'gl')())
  File "/usr/bin/gl", line 25, in importlib_load_entry_point
    return next(matches).load()
  File "/usr/lib/python3.10/importlib/metadata/__init__.py", line 162, in load
    module = import_module(match.group('module'))
  File "/usr/lib/python3.10/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/usr/lib/python3.10/site-packages/gitless/cli/gl.py", line 17, in <module>
    from . import (
  File "/usr/lib/python3.10/site-packages/gitless/cli/gl_fetch.py", line 26
    pprint.ok('Fetching from upstream succeeded')
    ^
IndentationError: expected an indented block after 'if' statement on line 25
```

**Comment by goldstar611** [comment]
> It doesn't seem to be working all the way.

I can't see all of the changes but you might need more spaces before that `pprint.ok` line. You'll need 4 spaces to make it part of the `if` statement.

**Comment by sirlucjan** [comment]
> > It doesn't seem to be working all the way.
> 
> I can't see all of the changes but you might need more spaces before that `pprint.ok` line. You'll need 4 spaces to make it part of the `if` statement.

Indeed. I gave too many spaces. 

Now I got another error.

```Traceback (most recent call last):
  File "/usr/bin/gl", line 33, in <module>
    sys.exit(load_entry_point('gitless==0.9.16', 'console_scripts', 'gl')())
  File "/usr/bin/gl", line 25, in importlib_load_entry_point
    return next(matches).load()
  File "/usr/lib/python3.10/importlib/metadata/__init__.py", line 162, in load
    module = import_module(match.group('module'))
  File "/usr/lib/python3.10/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/usr/lib/python3.10/site-packages/gitless/cli/gl.py", line 17, in <module>
    from . import (
  File "/usr/lib/python3.10/site-packages/gitless/cli/gl_fetch.py", line 25, in <module>
    if p.returncode == 0:
NameError: name 'p' is not defined
```

**Comment by sirlucjan** [comment]
I added a commit for the overview. At this hour (4 in the morning) a person is no longer thinking logically. Maybe I'll deal with this later.

**Comment by goldstar611** [comment]
> I added a commit for the overview. At this hour (4 in the morning) a person is no longer thinking logically. Maybe I'll deal with this later.

You were really, really close. I've pushed the fix and the tests back to your branch. I'll merge it shortly after the tests run.

```diff
diff --git a/gitless/cli/gl_fetch.py b/gitless/cli/gl_fetch.py
index d6a5a9f..3143f3a 100644
--- a/gitless/cli/gl_fetch.py
+++ b/gitless/cli/gl_fetch.py
@@ -22,7 +22,7 @@ def parser(subparsers, _):
 
 def main(args, repo):
     p = core.git_wrap('fetch', *args.fetch_args)
-if p.returncode == 0:
-    pprint.ok('Fetching from upstream succeeded')
-else:
-    pprint.err('Fetching from upstream failed')
+    if p.returncode == 0:
+        pprint.ok('Fetching from upstream succeeded')
+    else:
+        pprint.err('Fetching from upstream failed')
```

**Pull request merged**